libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -24,666 +24,850 @@
     }
 }
 
 /*
-  send_frames - sends updates down telemetry link for both DPORT and SPORT protocols
-  should be called by main program at 50hz to allow poll for serial bytes
-  coming from the receiver for the SPort protocol
-*/
-void AP_Frsky_Telem::send_frames(uint8_t control_mode)
+ * init - perform required initialisation
+ */
+void AP_Frsky_Telem::init(const AP_SerialManager &serial_manager, uint8_t *control_mode)
 {
-    // return immediately if not initialised
-    if (!_initialised_uart) {
-        return;
+    // check for protocol configured for a serial port - only the first serial port with one of these protocols will then run (cannot have FrSky on multiple serial ports)
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FRSky_DPort, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FRSky_DPort; // protocol for D-receivers
+    } else if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FRSky_SPort, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FRSky_SPort; // legacy Smart Port protocol (for X-receivers)
     }
 
-    if (_protocol == FrSkySPORT) {
-        if (!_mode_data_ready) {
-            _mode=control_mode;
-            _mode_data_ready = true;
-        }
-        if (!_baro_data_ready) {
-            calc_baro_alt();
-            _baro_data_ready = true;
-        }
-        if (!_gps_data_ready) {
-            calc_gps_position();
-            _gps_data_ready = true;
-        }
-        if (!_sats_data_ready) {
-            calc_gps_sats();
-            _sats_data_ready = true;
-        }
-        if (!_battery_data_ready) {
-            calc_battery();
-            _battery_data_ready = true;
-        }
-    } else {
-        _mode=control_mode;
-        send_hub_frame();
+    if (_port != NULL) {
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_Frsky_Telem::tick, void));
+        // we don't want flow control for either protocol
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
     }
 }
 
 /*
-  init_uart_for_sport - initialise uart for use by sport
-  this must be called from sport_tick which is called from the 1khz scheduler
-  because the UART begin must be called from the same thread as it is used from
+ * tick - main call to send data to the receiver (called by scheduler at 1kHz)
  */
-void AP_Frsky_Telem::init_uart_for_sport()
+void AP_Frsky_Telem::tick(void)
 {
-    // sanity check protocol
-    if (_protocol != FrSkySPORT) {
-        return;
+    static bool initialised_uart = false; // true when we have detected the protocol and UART has been initialised
+    // check UART has been initialised
+    if (!initialised_uart) {
+        // initialise uart (this must be called from within tick b/c the UART begin must be called from the same thread as it is used from)
+        if (_protocol == AP_SerialManager::SerialProtocol_FRSky_DPort) {     // for DPort protocol
```

Comment:
That's how it has been in the ArduPilot code all along. I'll make the necessary changes.